lib/luacompletion.c

```diff
@@ -0,0 +1,103 @@
+/*
+* SPDX-FileCopyrightText: (c) 2023-2024 Ring Zero Desenvolvimento de Software LTDA
+* SPDX-License-Identifier: MIT OR GPL-2.0-only
+*/
+
+#define pr_fmt(fmt) KBUILD_MODNAME ": " fmt
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/completion.h>
+
+#include <lua.h>
+#include <lualib.h>
+#include <lauxlib.h>
+
+#include <lunatik.h>
+
+typedef struct luacompletion_s {
+	struct completion event;
+} luacompletion_t;
+
+LUNATIK_OBJPRIVCHECKER(luacompletion_check, luacompletion_t *);
+
+static int luacompletion_complete(lua_State *L)
+{
+	luacompletion_t *completion = luacompletion_check(L, 1);
+
+	complete(&completion->event);
+	return 0;
+}
+
+static int luacompletion_wait(lua_State *L)
+{
+	luacompletion_t *completion = luacompletion_check(L, 1);
+
+	wait_for_completion(&completion->event);
+	return 0;
+}
+
+static int luacompletion_wait_timeout(lua_State *L)
+{
+	luacompletion_t *completion = luacompletion_check(L, 1);
+	unsigned long timeout = (unsigned long)luaL_checkinteger(L, 2);
+	unsigned long ret;
+
+	ret = wait_for_completion_timeout(&completion->event, msecs_to_jiffies(timeout));
+	lua_pushboolean(L, ret > 0 ? true : false);
+	return 1;
+}
+
+
+static void luacompletion_release(void *private)
+{
+	luacompletion_t *completion = (luacompletion_t *)private;
+	(void)completion; /* do nothing */
```

Comment:
If we have no destructor, you can just avoid this function completely.. e.g., `.release = NULL`